File: leetcode/0142-linked-list-cycle-II/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * RUNTIME: 17 ms (60.40%)
     * MEMORY: 9.2 MB (22.16%)
    */
    ListNode *detectCycle(ListNode *head) {
        if (!head || !head->next || !head->next->next) {
            return NULL;
        }
        unordered_map <ListNode*, int> slowTraversed;
        slowTraversed[head] = 0;
        ListNode* slow = head->next;
        ListNode* fast = head->next->next;
        while (slow != fast) {
            slowTraversed[slow] = 0;
            if (!fast->next || !fast->next->next) {
                return NULL;
            }
            if (slowTraversed.find(fast) != slowTraversed.end()) {
                return fast;
            }
            else if (slowTraversed.find(fast->next) != slowTraversed.end()) {
                return fast->next;
            }
            if (slowTraversed.find(fast->next->next) != slowTraversed.end()) {
                return fast->next->next;
            }
            slow = slow->next;
            fast = fast->next->next;
        }
        return fast;
    }
};

File: leetcode/0142-linked-list-cycle-II/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        RUNTIME: 52 ms (93.93%)
        MEMORY: 17.9 MB (23.21%)
        """
        if not head or not head.next or not head.next.next: return None
        slow_traversed = {}
        slow_traversed[head] = 0
        slow, fast = head.next, head.next.next
        while slow != fast:
            slow_traversed[slow] = 0
            if not fast.next or not fast.next.next: return None
            if fast in slow_traversed:
                return fast
            elif fast.next in slow_traversed:
                return fast.next
            elif fast.next.next in slow_traversed:
                return fast.next.next
            slow = slow.next
            fast = fast.next.next
        return fast
